receiver/receivercreator/discovery.go

```diff
@@ -0,0 +1,183 @@
+// Scraper configurations are only created for Port Endpoints.
+// TODO: Log receiver configurations are only created for Pod Container Endpoints.
+func (builder *k8sHintsBuilder) createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var pod observer.Pod
+
+	endpointType := getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if endpointPod, ok := env["pod"]; ok {
+		err := mapstructure.Decode(endpointPod, &pod)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+	} else {
+		return nil, nil
+	}
+
+	return builder.createScraper(pod.Annotations, env)
+}
+
+func (builder *k8sHintsBuilder) createScraper(
+	annotations map[string]string,
+	env observer.EndpointEnv) (*receiverTemplate, error) {
+	var port uint16
+	var p observer.Port
+	err := mapstructure.Decode(env, &p)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract port event: %v", zap.Any("env", env))
+	}
+	if p.Port == 0 {
+		return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+	}
+	port = p.Port
+	pod := p.Pod
+
+	if !discoveryMetricsEnabled(annotations, otelMetricsHints, fmt.Sprint(port)) {
+		return nil, nil
+	}
+
+	subreceiverKey, found := getHintAnnotation(annotations, otelMetricsHints, scraperHint, fmt.Sprint(port))
+	if !found || subreceiverKey == "" {
+		// no scraper hint detected
+		return nil, nil
+	}
+	if _, ok := builder.ignoreReceivers[subreceiverKey]; ok {
+		// scraper is ignored
+		return nil, nil
+	}
+	builder.logger.Debug("handling added hinted receiver", zap.Any("subreceiverKey", subreceiverKey))
+
+	defaultEndpoint := getStringEnv(env, endpointConfigKey)
+	userConfMap := getScraperConfFromAnnotations(annotations, defaultEndpoint, fmt.Sprint(port), builder.logger)
+
+	recTemplate, err := newReceiverTemplate(fmt.Sprintf("%v/%v_%v", subreceiverKey, pod.UID, port), userConfMap)
+	recTemplate.signals = receiverSignals{true, false, false}
+
+	return &recTemplate, err
+}
+
+func getScraperConfFromAnnotations(
+	annotations map[string]string,
+	defaultEndpoint, scopeSuffix string,
+	logger *zap.Logger) userConfigMap {
+	conf := userConfigMap{}
+	conf[endpointConfigKey] = defaultEndpoint
+
+	configStr, found := getHintAnnotation(annotations, otelMetricsHints, configHint, scopeSuffix)
+	if !found || configStr == "" {
+		return conf
+	}
+	if err := yaml.Unmarshal([]byte(configStr), &conf); err != nil {
+		logger.Debug("could not unmarshal configuration from hint", zap.Error(err))
+	}
+
+	val := conf[endpointConfigKey]
+	confEndpoint, ok := val.(string)
+	if !ok {
+		logger.Debug("could not extract configured enpoint")
+		return userConfigMap{}
+	}
+
+	if !strings.Contains(confEndpoint, defaultEndpoint) {
```

Comment:
I have a couple of questions here:

1. Will this work for annotation with "`endpoint`" string in it? like the one we have in the readme
```
io.opentelemetry.discovery.metrics/config: |
  endpoint: "http://`endpoint`/nginx_status"
```

2. I think "contains" isn't a strong enough check. Someone can configure the annotation to be 
"http://another-endpoint/nginx_status?my-endpoint:8888" and if collector will scrape "another-endpoint". I think we should use URI pattern check here instead

Reply:
Right to both. The validation had to be improved indeed. Fixed that in the last commit.